File: backend/app/schemas/gamification.py
from datetime import datetime

from app.schemas.base import CamelModel


class AchievementOut(CamelModel):
    key: str
    name: str
    description: str
    earned_at: datetime | None = None  # None only in the "locked" list


class GamificationSummary(CamelModel):
    xp: int
    level: int
    xp_into_level: int
    xp_to_next_level: int
    earned_achievements: list[AchievementOut]
    locked_achievements: list[AchievementOut]


class LeaderboardEntryOut(CamelModel):
    rank: int
    student_id: str
    name: str
    xp: int
    level: int
    is_current_student: bool

File: src/types/gamification.ts
export interface Achievement {
  key: string;
  name: string;
  description: string;
  earnedAt: string | null; // ISO datetime, null only for locked achievements
}

export interface GamificationSummary {
  xp: number;
  level: number;
  xpIntoLevel: number;
  xpToNextLevel: number;
  earnedAchievements: Achievement[];
  lockedAchievements: Achievement[];
}

export type LeaderboardScope = 'class' | 'all';

export interface LeaderboardEntry {
  rank: number;
  studentId: string;
  name: string;
  xp: number;
  level: number;
  isCurrentStudent: boolean;
}
